src/rhsmlib/facts/pkg_arches.py

```diff
@@ -0,0 +1,74 @@
+# -*- coding: utf-8 -*-
+
+from __future__ import print_function, division, absolute_import
+
+#
+# Copyright (c) 2020 ATIX AG
+#
+# This software is licensed to you under the GNU General Public License,
+# version 2 (GPLv2). There is NO WARRANTY for this software, express or
+# implied, including the implied warranties of MERCHANTABILITY or FITNESS
+# FOR A PARTICULAR PURPOSE. You should have received a copy of GPLv2
+# along with this software; if not, see
+# http://www.gnu.org/licenses/old-licenses/gpl-2.0.txt.
+#
+# Red Hat trademarks are not licensed under GPLv2. No permission is
+# granted to use or replicate Red Hat trademarks that are incorporated
+# in this software or its documentation.
+#
+import logging
+import subprocess
+
+from rhsmlib.facts import collector
+
+log = logging.getLogger(__name__)
+
+
+class SupportedArchesCollector(collector.FactsCollector):
+    """
+    Class used for collecting packages architectures of a host
+    """
+
+    def __init__(self, arch=None, prefix=None, testing=None, collected_hw_info=None):
+        super(SupportedArchesCollector, self).__init__(
+            arch=arch,
+            prefix=prefix,
+            testing=testing,
+            collected_hw_info=collected_hw_info)
+
+    def get_arches_on_debuntu(self):
+        """
+        Try to return content of all supported packages architectures
+        :return: dictionary containing architectures
+        Otherwise empty dictionary is returned.
+        """
+        arches = []
+
+        try:
+            arch = subprocess.check_output("dpkg --print-architecture", shell=True).decode('UTF-8')
+            if arch != '':
+                arches.append(arch.rstrip('\n'))
+        except Exception as e:
+            log.error("Error getting dpkg main architecture: %s", e)
+
+        try:
+            arch = subprocess.check_output("dpkg --print-foreign-architectures", shell=True).decode('UTF-8')
+            if arch != '':
+                arches.append(arch.rstrip('\n'))
+        except Exception as e:
+            log.error("Error getting dpkg foreign architecture: %s", e)
+
+        return {
+            'supported_architectures': ','.join(arches)
+        }
+
+    def get_all(self):
+        """
+        Get all architectures of a debian / ubuntu host
+        :return: dictionary containing architectures
+        """
+        dist_name = self._collected_hw_info['distribution.name']
+        dist_name = dist_name.lower()
+        if 'ubuntu' in dist_name or 'debian' in dist_name:
```

Comment:
better make the list of Debian architectures as proper list, possibly as class constant, so it is easier to add support for a new Debian-based distribution in the future [1]

[1] or, even better, extending `HardwareCollector` to read `ID_LIKE` from `/etc/os-release` as `'distribution.id_like` fact, and use it as additional way to support any Debian-based distribution

Reply:
I like your suggestion to use `ID_LIKE`. But AFAIK the `ID_LIKE` parameter is not always given. Therefore, we cannot rely on it in every case and I would suggest the following:
Read both (`'distribution.name'` + `'distribution.id_like'`) and check whether at least one of these is part of the supported OS list.

What do you think about it?

Reply:
It was more a comment for some future improvement :-)

Oh, and I just noticed an important thing to change: `distribution.name` is read from `NAME` in `os-release`, which is a user-facing name of the distribution! We really need to use `ID`, so `distribution.id`. See [1].

[1] https://www.freedesktop.org/software/systemd/man/os-release.html

In the short term, IMHO a good solution is something like:
```python
    DEBIAN_DISTRIBUTIONS = ['debian', 'ubuntu']
...
    dist_id = self._collected_hw_info['distribution.id']
    if dist_id in DEBIAN_DISTRIBUTIONS:
        do_something_on_debian_based_distributions()
```
So:
- it covers both Debian and Ubuntu
- it is easy to add support another Debian-based distribution (append a new value to `DEBIAN_DISTRIBUTIONS`)
- in case we read `ID_LIKE` from `os-release` as fact, checking for `debian` as `distribution.name` or in `distribution.id_like` will be the first check, leaving the code above as fallback for very old systems (or even drop it, since `/etc/os-release` has been shipped in Debian for many years)

Reply:
The check must be the other way around: I want to check whether `ubuntu` or `debian` is part of `dist_name` since it could be `debian gnu/linux`. With your suggestion `dist_name` must contain exactly `debian` or `ubuntu`. Therefore, we would need a for loop at this point.

Reply:
See what I wrote above about `distribution.name` vs `distribution.id` (which is what we need to use).

Reply:
Yeah and it makes sense to use it. Sry, I should have said `dist_id` before. 

But the same problem applies to `distribution.id`. The `ID` could contain something else (I have an openSUSE machine running where the `ID` states `opensuse-tumbleweed` and not just `opensuse`) so we need to check whether any of the supported operating systems contains a substring of the `dist_id`. The code must be like the following then:

```
DEBIAN_DISTRIBUTIONS = ['debian', 'ubuntu']
...
dist_id = self._collected_hw_info['distribution.id']
if any(os in dist_id for os in DEBIAN_DISTRIBUTIONS):
    do_something_on_debian_based_distributions()
```

Do you agree?

Reply:
openSUSE has different IDs depending on whether it is a released version, it is Tumbleweed, etc. I'd not worry about that unless we need to check for openSUSE using the `distribution` facts.

This code is taylored for Debian and Debian-based distributions, so an exact string match on the `ID` will work, as it is exactly what the `ID` is for. Debian has `debian`, Ubuntu has `ubuntu`, and most probably any other derived from Debian has either `debian` or `ubuntu` (or both) in `ID_LIKE` (this is something to worry only when we have `ID_LIKE` as fact).

Reply:
I just tested it on a Debian Buster with `distribution.id`. Unfortunately, the `HardwareCollector` returns the `VERSION` field of `/etc/os-release` as `distribution.id` (which is kind of misleading).
Either a bug or works-as-designed [Line 161 HardwareCollector](https://github.com/candlepin/subscription-manager/blob/master/src/rhsmlib/facts/hwprobe.py#L161). What do you think?

Therefore, I would stick to `distribution.name` in this PR.

Reply:
Urghhh that sucks so much, thanks for pointing that out. Then I guess your proposed code (the one with `any()`) is what is needed, using a lowercased `distribution.name`:
```python
DEBIAN_DISTRIBUTIONS = ['debian', 'ubuntu']
...
dist_name = self._collected_hw_info['distribution.name']
if any(os in dist_name for os in DEBIAN_DISTRIBUTIONS):
    do_something_on_debian_based_distributions()
```

Reply:
Implemented as suggested.